File: TimMovie/TimMovie.Web/Controllers/AccountController.cs
using System.Security.Claims;
using AutoMapper;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Identity;
using Microsoft.AspNetCore.Mvc;
using TimMovie.Core.DTO.Account;
using TimMovie.Core.Entities;
using TimMovie.Core.Interfaces;
using TimMovie.Core.Services.Banners;
using TimMovie.Web.ViewModels;
using TimMovie.Web.ViewModels.Account;

namespace TimMovie.Web.Controllers;

[AllowAnonymous]
public class AccountController : Controller
{
    private readonly ILogger<AccountController> logger;
    private readonly UserManager<User> userManager;
    private readonly IMapper mapper;
    private readonly IUserService userService;
    private readonly BannerService bannerService;

    public AccountController(UserManager<User> userManager,
        IMapper mapper,
        ILogger<AccountController> logger,
        IUserService userService)
    {
        this.userManager = userManager;
        this.mapper = mapper;
        this.logger = logger;
        this.userService = userService;
    }

    [HttpGet]
    public IActionResult Registration(string? returnUrl)
    {
        return View();
    }

    [HttpPost]
    public async Task<IActionResult> RegistrationAsync(RegistrationViewModel model)
    {
        if (!ModelState.IsValid)
        {
            return View(model);
        }

        var userDto = mapper.Map<UserRegistrationDto>(model);
        var registerUserResult = await userService.RegisterUserAsync(userDto);

        if (!registerUserResult.Succeeded)
        {
            AddErrors(registerUserResult);
            return View(model);
        }

        return View();
    }

    [HttpPost]
    public async Task<IActionResult> LoginAsync(string login, string password, bool rememberMe)
    {
        var loginDto = new LoginDto
        {
            Login = login,
            Password = password,
            RememberMe = rememberMe
        };


        var loginResult = await userService.LoginAsync(loginDto);

        if (loginResult.Succeeded)
        {
            logger.LogInformation($"пользователь {login} вошел в свой акканут");
            return Ok();
        }

        if (loginResult.IsNotAllowed)
        {
            var userFromDb = await userManager.FindByNameAsync(login) ?? await userManager.FindByEmailAsync(login);
            return PartialView("MailSend", (userFromDb.Email, userFromDb.DisplayName));
        }

        logger.LogInformation($"неудачная попытка входа с использованием логина {login}");
        return Ok("Неверный(е) логин/почта или пароль");
    }

    [HttpGet]
    public IActionResult Denied()
    {
        return View("Denied");
    }

    private void AddErrors(IdentityResult result)
    {
        foreach (var error in result.Errors)
        {
            ModelState.AddModelError(string.Empty, error.Description);
        }
    }
}

File: TimMovie/TimMovie.WebApi/Controllers/AccountController.fs
namespace TimMovie.WebApi.Controllers.AuthorizationController

open Microsoft.AspNetCore.Authorization
open Microsoft.AspNetCore.Http
open Microsoft.AspNetCore.Mvc
open Microsoft.Extensions.Logging
open TimMovie.Core.DTO.Account
open TimMovie.Core.Interfaces
open TimMovie.SharedKernel.Classes

[<ApiController>]
[<Route("[controller]/[action]")>]
type AccountController(logger: ILogger<AccountController>, userService: IUserService) as this =
    inherit ControllerBase()

    member private _.UrlToConfirmEmail =
        this.Url.Action("ConfirmEmail", "Account", null, this.HttpContext.Request.Scheme)

    [<HttpPost>]
    [<AllowAnonymous>]
    [<Consumes("application/x-www-form-urlencoded")>]
    member _.Register([<FromForm>] email: string, [<FromForm>] username: string, [<FromForm>] password: string) =
        let userRegistrationDto = UserRegistrationDto()
        userRegistrationDto.Email <- email
        userRegistrationDto.UserName <- username
        userRegistrationDto.Password <- password
        
        let registerResult =
            userService.RegisterUserAsync(userRegistrationDto)
            |> Async.AwaitTask
            |> Async.RunSynchronously

        let result =
            if registerResult.Succeeded then
                userService.SendConfirmEmailAsync(email, this.UrlToConfirmEmail)
                |> Async.AwaitTask
                |> Async.RunSynchronously
            else
                let sb = System.Text.StringBuilder()
                for error in registerResult.Errors do
                    sb.Append($" {error.Description}") |> ignore

                Result.Fail($"{sb.ToString()}")

        result

    [<HttpPost>]
    [<AllowAnonymous>]
    [<Consumes("application/x-www-form-urlencoded")>]
    member _.SendConfirmEmail([<FromForm>] email: string) =
        userService.SendConfirmEmailAsync(email, this.UrlToConfirmEmail)

    [<HttpGet>]
    [<AllowAnonymous>]
    [<ApiExplorerSettings(IgnoreApi=true)>]
    member _.ConfirmEmail(userId: string, code: string) =
        userService.ConfirmEmailAsync(userId, code)
